src/NHibernate/Util/AsyncLock.cs

```diff
@@ -0,0 +1,47 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace NHibernate.Util
+{
+	public class AsyncLock
+	{
+		private readonly AsyncSemaphore semaphore;
+		private readonly Task<Releaser> releaser;
+
+		public AsyncLock()
+		{
+			semaphore = new AsyncSemaphore(1);
+			releaser = Task.FromResult(new Releaser(this));
+		}
+
+		public Task<Releaser> LockAsync()
+		{
+			var wait = semaphore.WaitAsync();
+			return wait.IsCompleted
+				? releaser
+				: wait.ContinueWith(
+					(_, state) => new Releaser((AsyncLock)state),
```

Comment:
Couldn't this be simplified to `new Releaser(this)`? Going through `state` obfuscates the code in my opinion.